### ISSUE TITLE
Unexpected error assigning with unions

### ISSUE THREAD

**Issue by molinger** [microsoft/TypeScript#51182]
# Bug Report

### 🔎 Search Terms

union not assignable

I found a couple bugs that seemed related in the title but after reviewing, they appeared to be separate issues. They also reported different behaviors in older versions but this bug is consistent in those versions.

https://github.com/microsoft/TypeScript/issues/50488
https://github.com/microsoft/TypeScript/issues/51180

### 🕗 Version & Regression Information

- I was unable to test this on prior versions because `Common "Bugs" That Aren't Bugs`

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.8.4&ssl=42&ssc=2&pln=1&pc=1#code/JYOwLgpgTgZghgYwgAgKIFs7ADYEkAmE4wMw0yA3gFDK3JgCeADhAFzIBEGW2HANDTpx8+KBADO49hwBWAewAWIAHT45EAAIQAHnHRNsEZQjnoOVAL5UqoSLEQoAcgFdseQsVLlqdeszacLm7mVjbg0PBIaJg4AMJy4dpglIK0cmAK0OwgzugARtAA3KnIwB5gJGRQ7Nw4BEQVXlDFobYRDshB2PGJyT506ZnVyDn5RSVlDZVZna7uU00t1jDOIAgVCcgKcCD4hj2QSQAUJr01Md0Jh8kAPrNuBzpgAJQplsur68CbAOYQYPVPFUjs9zjxAY0qsg7l0IdMoClfGIwM4oCBKH4WNIuhxkBYllQVmsNuiMsBxAAJODiVBQKByKAgxF0ba7fZXJ5Hfq+NIZGYABgEPLokyBMz+APK8JBJQszwJRK+mzJ4gA6gyANbiJnc5CncTJUWQ8gAXmQErhTRBxV8JGQRyN8OUjBYyBN7sCcw4r11vlZewgj2OFGQgwFfFKUqaePlsuQEGw4hQvpZOwDQbAXNDfOGgsjCyhcptdCsFiAA)

### 💻 Code

```ts
interface EmailIdentifier {
    type: "Email",
    address: "john.doe@example.com"
}

interface NullIdentifier {
    type: "Null"
}

interface EmailContext {
    other: number;
    identifier: EmailIdentifier;
}

interface NullContext {
    other: number;
    identifier: NullIdentifier;
}

function handleContext(context: EmailContext | NullContext) {

}

function getIdentifier(): EmailIdentifier | NullIdentifier {
    return { type: "Null" };
}

function thisHasError() {
    handleContext({
        other: 0,
        identifier: getIdentifier()
    });
}

function thisWorks() {
    const identifier = getIdentifier();
    if (identifier.type === "Null") {
        handleContext({ other: 0, identifier });
    } else {
        handleContext({ other: 0, identifier });
    }
}
```

### 🙁 Actual behavior

Errors in code

    Argument of type '{ other: number; identifier: EmailIdentifier | NullIdentifier; }' is not assignable to parameter of type 'EmailContext | NullContext'.
      Type '{ other: number; identifier: EmailIdentifier | NullIdentifier; }' is not assignable to type 'NullContext'.
        Types of property 'identifier' are incompatible.
          Type 'EmailIdentifier | NullIdentifier' is not assignable to type 'NullIdentifier'.
            Type 'EmailIdentifier' is not assignable to type 'NullIdentifier'.
              Types of property 'type' are incompatible.
                Type '"Email"' is not assignable to type '"Null"'.
    'context' is declared but its value is never read.

It isn't clear why we go from assigning to the type `EmailContext | NullContext` to just `NullContext`.

### 🙂 Expected behavior

No error.

**Comment by MartinJohns** [comment]
> It isn't clear why we go from assigning to the type EmailContext | NullContext to just NullContext.

In your `thisHasError()` case the inferred type is:

```typescript
{ other: number; identifier: EmailIdentifier | NullIdentifier }
```

This is not compatible with `{ other: number; identifier: EmailIdentifier } | { other: number; identifier: NullIdentifier }`.

Whereas in your working case the types are respectively:
```typescript
{ other: number; identifier: EmailIdentifier }
// and
{ other: number; identifier: NullIdentifier }
```

**Comment by RyanCavanaugh** [comment]
*In general*, a `{ x: T | U } ` is not safely assignable to a `{ x: T } | { x: U }`, because as soon as there's more than one `x` property, new combinations like `{ x1: T, x2: U }` go outside the domain of the target (`{ x1: T, x2: T } | { x1: U, x2: U }`). There are some places where TS is willing to iterate through a union, but in this case since it's potentially multiplicatively explosive, it doesn't.

[Thread with more context](https://twitter.com/SeaRyanC/status/1544414378383925250)

**Comment by molinger** [comment]
Thanks for the clarification.
